base/src/commonMain/kotlin/state/State.kt

```diff
@@ -192,35 +190,35 @@ abstract class StateRepo<S:State<S>>(coroutineContext: CoroutineContext = MainQu
      * If the [action] returns [State.remain] no state transition will occur.
      * Since this operation is atomic, the [action] should not directly call [takeAndChangeState] itself. If required to do this, handle the additional transition in a separate coroutine.
      *
+     * This method uses a separate coroutineScope, meaning it will suspend until all child Jobs are completed, including those that asynchronously call this method itself.
+     *
      * @param action Function to determine the [State] to be transitioned to from the current [State]. If no state transition should occur, return [State.remain]
      */
-    suspend fun takeAndChangeState(action: suspend (S) -> suspend () -> S): S {
+    suspend fun takeAndChangeState(action: suspend (S) -> suspend () -> S): S = coroutineScope { // scope around the mutex so asynchronously scheduled coroutines that also use this method can run before the scope completed without deadlocks
```

Comment:
Btw @Daeda88 I did re-add a coroutine scope, but putting this around the Mutex ensures we don't get the deadlocks as before.

It helps writing tests (and I imagine code) to know all async work will be completed before the  scope ends. To be fair some tests to verify this would be great, but I ran out of time.